python/semantic_kernel/connectors/ai/function_choice_behavior.py

```diff
@@ -0,0 +1,244 @@
+
+from pydantic.dataclasses import dataclass
+
+from semantic_kernel.exceptions.service_exceptions import ServiceInvalidExecutionSettingsError
+from semantic_kernel.functions.kernel_function_metadata import KernelFunctionMetadata
+from semantic_kernel.kernel_pydantic import KernelBaseModel
+from semantic_kernel.utils.experimental_decorator import experimental_class, experimental_function
+
+if TYPE_CHECKING:
+    from semantic_kernel.connectors.ai.function_call_behavior import FunctionCallBehavior
+    from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
+    from semantic_kernel.kernel import Kernel
+
+DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS = 5
+
+logger = logging.getLogger(__name__)
+
+
+@experimental_class
+class FunctionChoiceType(Enum):
+    """The type of function choice behavior."""
+
+    AUTO = "auto"
+    NONE = "none"
+    REQUIRED = "required"
+
+
+@experimental_function
+def _check_for_missing_functions(function_names: list[str], kernel_function_metadata: set) -> None:
+    """Check for missing functions in the kernel function metadata."""
+    try:
+        qualified_names = {metadata.fully_qualified_name for metadata in kernel_function_metadata}
+        missing_functions = [name for name in function_names if name not in qualified_names]
+    except Exception:
+        logger.exception("Error checking for missing functions.")
+        return
+
+    if missing_functions:
+        raise ServiceInvalidExecutionSettingsError(
+            f"Unable to find the required function(s) `{', '.join(missing_functions)}` "
+            f"in the list of available kernel functions: {', '.join(qualified_names)}."
+        )
+
+
+@experimental_class
+@dataclass
+class FunctionCallChoiceConfiguration:
+    available_functions: list["KernelFunctionMetadata"] | None = None
+
+
+@experimental_class
+class FunctionChoiceBehavior(KernelBaseModel):
+    """Class that controls function choice behavior.
+
+    Attributes:
+        enable_kernel_functions: Enable kernel functions.
+        max_auto_invoke_attempts: The maximum number of auto invoke attempts.
+        filters: Filters for the function choice behavior.
+        function_fully_qualified_names: The fully qualified names of the functions.
+        type: The type of function choice behavior.
+
+    Properties:
+        auto_invoke_kernel_functions: Check if the kernel functions should be auto-invoked.
+            Determined as max_auto_invoke_attempts > 0.
+
+    Methods:
+        configure: Configures the settings for the function call behavior,
+            the default version in this class, does nothing, use subclasses for different behaviors.
+
+    Class methods:
+        Auto: Returns FunctionChoiceAuto class with auto_invoke enabled, and the desired functions
+            based on either the specified filters or the full qualified names.
+        NoneInvoke: Returns FunctionChoiceNone class with auto_invoke disabled, and the desired functions
```

Comment:
so this is non-auto invoke basically, right? would be good to better document, because it is not immediately obvious to me...